contracts/router/RouterV2.sol

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {Ownable} from "@openzeppelin/contracts-4.5.0/access/Ownable.sol";
+
+import {DefaultRouter} from "./DefaultRouter.sol";
+import {BridgeFailed, ModuleExists, ModuleNotExists, QueryEmpty} from "./libs/Errors.sol";
+import {Action, BridgeToken, DestRequest, LimitedToken, SwapQuery} from "./libs/Structs.sol";
+
+import {ISwapQuoterV1} from "./interfaces/ISwapQuoterV1.sol";
+import {IBridgeModule} from "./interfaces/IBridgeModule.sol";
+import {IRouterV2} from "./interfaces/IRouterV2.sol";
+
+contract RouterV2 is IRouterV2, DefaultRouter, Ownable {
+    /// @notice swap quoter
+    ISwapQuoterV1 public immutable swapQuoter;
+
+    /// @notice List of all connected bridge modules
+    address[] internal _bridgeModules;
+
+    /// @inheritdoc IRouterV2
+    mapping(bytes32 => address) public idToModule;
+
+    /// @inheritdoc IRouterV2
+    mapping(address => bytes32) public moduleToId;
+
+    event ModuleConnected(bytes32 moduleId, address indexed bridgeModule);
+
+    constructor(address _swapQuoter) {
+        swapQuoter = ISwapQuoterV1(_swapQuoter);
+    }
+
+    /// @inheritdoc IRouterV2
+    function bridgeViaSynapse(
+        address to,
+        uint256 chainId,
+        bytes32 moduleId,
+        address token,
+        uint256 amount,
+        SwapQuery memory originQuery,
+        SwapQuery memory destQuery
+    ) external payable {
+        address bridgeModule = idToModule[moduleId];
+        if (bridgeModule == address(0)) revert ModuleNotExists();
+
+        // pull (and possibly swap) token into router
+        if (_hasAdapter(originQuery)) {
+            (token, amount) = _doSwap(address(this), token, amount, originQuery);
+        } else {
+            _pullToken(address(this), token, amount);
+        }
+
+        // delegate bridge call to module
+        // @dev delegatecall should approve to spend
+        bytes memory payload = abi.encodeWithSelector(
+            IBridgeModule.delegateBridge.selector,
+            to,
+            chainId,
+            token,
+            amount,
+            destQuery
+        );
+        (bool success, ) = bridgeModule.delegatecall(payload);
```

Comment:
Let's bubble up the revert message here:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5c4b75a9cdda99af901b8c153d0d1e8a90b0139a/contracts/utils/Address.sol#L174C14-L174C34

Reply:
Um, just realized that in `LinkedPool` the revert message is also not bubbled. Might need to consider redeploying those.